doc/sphinxext/redirect_from.py

```diff
@@ -0,0 +1,84 @@
+"""
+Redirecting old docs to new location
+====================================
+
+If an rst file is moved or its content subsumed in a different file, it
+is desireable to redirect the old file to the new or existing file. This
+extension enables this with a simple html refresh.
+
+For example suppose ``doc/topic/old-page.rst`` is removed and its content
+included in ``doc/topic/new-page.rst``.  We use the ``redirect_from``
+directive in ``doc/topic/new-page.rst``::
+
+    .. redirect-from:: /topic/old-page
+
+This creates in the build directory a file ``build/html/topic/old-page.html``
+that contains a relative refresh::
+
+    <html>
+      <head>
+        <meta http-equiv="refresh" content="0; url=new-page.html">
+        <link rel="canonical" href="new-page.html">
+      </head>
+    </html>
+
+If you need to redirect across subdirectory trees, that works as well.  For
+instance if ``doc/topic/subdir1/old-page.rst`` is now found at
+``doc/topic/subdir2/new-page.rst`` then ``new-page.rst`` just lists the
+full path::
+
+    .. redirect-from:: /topic/subdir1/old-page.rst
+
+"""
+
+from pathlib import Path
+from docutils.parsers.rst import Directive
+from sphinx.util import logging
+
+logger = logging.getLogger(__name__)
+
+
+HTML_TEMPLATE = """<html>
+  <head>
+    <meta http-equiv="refresh" content="0; url={v}">
+    <link rel="canonical" href="{v}">
```

Comment:
Is this going to work without being a full URL? Will we need to post-process this for upload to the docs site?

Reply:
Hmmm good point.  This says you can use relative URLs https://stackoverflow.com/questions/8467938/do-canonical-links-require-a-full-domain  But Google seems to want a full name: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls?visit_id=637484795093592518-2178287391&hl=en&rd=2#expandable-7

We could just add `https://matplotlib.org/stable` to this string.  The script we are proposing in the GitHub page could clean up the canonical later if needed.

Reply:
Actually lets just drop this.  95% sure search engines don't read the redirect anyway...